sw/host/tests/chip/spi_device/src/spi_device_sleep_test.rs

```diff
@@ -0,0 +1,112 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{bail, Result};
+use clap::Parser;
+use regex::Regex;
+use std::time::Duration;
+
+use opentitanlib::app::TransportWrapper;
+use opentitanlib::execute_test;
+use opentitanlib::io::gpio::{GpioPin, PinMode, PullMode};
+use opentitanlib::io::spi::Transfer;
+use opentitanlib::io::uart::Uart;
+use opentitanlib::spiflash::SpiFlash;
+use opentitanlib::test_utils::init::InitializeTest;
+use opentitanlib::uart::console::{ExitStatus, UartConsole};
+
+#[derive(Debug, Parser)]
+struct Opts {
+    #[command(flatten)]
+    init: InitializeTest,
+
+    /// Console receive timeout.
+    #[arg(long, value_parser = humantime::parse_duration, default_value = "600s")]
+    timeout: Duration,
+
+    /// Name of the SPI interface to connect to the OTTF console.
+    #[arg(long, default_value = "BOOTSTRAP")]
+    spi: String,
+}
+
+fn spi_sleep_test(opts: &Opts, transport: &TransportWrapper) -> Result<()> {
+    let uart = transport.uart("console")?;
+    let spi = transport.spi(&opts.spi)?;
+    let wakeup_pin = transport.gpio_pin("IOA8")?;
+    wakeup_pin.set_mode(PinMode::PushPull)?;
+    let sleep_indicator = transport.gpio_pin("IOA7")?;
+    sleep_indicator.set_mode(PinMode::Input)?;
+    sleep_indicator.set_pull_mode(PullMode::PullUp)?;
+
+    const DATA_SIZE: usize = 10;
+
+    wait_sleep(&*sleep_indicator, &*uart, opts.timeout)?;
+    /* Test the spi dev in sleep with the SD0 in high. */
+    let mut buf = [0x55u8; DATA_SIZE];
+    spi.run_transaction(&mut [Transfer::Write(&[0x55; 1]), Transfer::Read(&mut buf)])?;
+    assert_eq!(&buf, &[0xffu8; DATA_SIZE]);
+    wake_up(&*wakeup_pin, &*uart, opts.timeout)?;
+
+    wait_sleep(&*sleep_indicator, &*uart, opts.timeout)?;
+    /* Test the spi dev in sleep with the SD0 in low. */
+    let mut buf = [0x55u8; DATA_SIZE];
+    spi.run_transaction(&mut [Transfer::Write(&[0x55; 1]), Transfer::Read(&mut buf)])?;
+    assert_eq!(&buf, &[0x00u8; DATA_SIZE]);
+    wake_up(&*wakeup_pin, &*uart, opts.timeout)?;
+
+    Ok(())
+}
+
+fn spi_wakeup_test(opts: &Opts, transport: &TransportWrapper) -> Result<()> {
+    let sleep_indicator = transport.gpio_pin("IOA7")?;
+    sleep_indicator.set_mode(PinMode::Input)?;
+    sleep_indicator.set_pull_mode(PullMode::PullUp)?;
+    let uart = transport.uart("console")?;
+    let spi = transport.spi(&opts.spi)?;
+    spi.set_max_speed(100_000)?;
+
+    wait_sleep(&*sleep_indicator, &*uart, opts.timeout)?;
+    println!("Reading jedec id");
+    // Should wakup when the cs goes low.
+    let jdec_id = SpiFlash::read_jedec_id(&*spi, 5)?;
+    assert_eq!(&jdec_id, &[0x17u8, 0x17, 0x74, 0x98, 0x22]);
```

Comment:
This transaction is probably failing because it sets too unreasonable a burden on OT to wake up and clear sleep mode before the read data must start being returned.

A transaction that is used to wake up the chip would need to be discarded. If you just want CSB to wake the chip up, then you can probably do something like...

```rust
    // Wake up chip with CSB. Hold asserted until the chip releases sleep mode.
    let csb_hold = spi.assert_cs()?;
    wait_awake(); // Wait for IOA7 to return to high.
    drop(csb_hold);

    // Check that transactions now work.
    let jdec_id = SpiFlash::read_jedec_id(&*spi, 5)?;
    assert_eq!(&jdec_id, &[0x17u8, 0x17, 0x74, 0x98, 0x22]);
```

Reply:
Should be fixed now by the configuration of the spi_device.